### PR TITLE
Recharge Model Features - Peterson & Western / Flex (Groundwater Uptake)

### PR DESCRIPTION
**Pull request by martinvonk**
# Description
This pull request introduces two new features to the recharge models:

1.  Peterson & Western nonlinear recharge model. This model was originally described by [Peterson & Western (2014) - Nonlinear time-series modeling of unconfined groundwater head](http://doi.wiley.com/10.1002/2013WR014800). The recharge model is based on equation 6 from this article and has five parameters. If parameter `gamma = 1`, the model behaves similarly to the FlexModel which is already implemented in Pastas. Some good time series model results were obtained with this recharge model, compared to other nonlinear recharge models.
2. A root uptake parameter for the `gw_uptake=True` optionality in the FlexModel is added. Currently the excess evaporation is subtracted fully from the recharge flux when `gw_uptake=True`. With this pull request a parameter `f` is introduced which scales the excess evaporation between 0 and 1. This parameter is found to take sensible values, while not changing other parameter values too much.

I have not made tests for these features yet (don't know how :)).
I'll add an example notebook to the repository [pastas/test-datasets](https://github.com/pastas/test-datasets) because we have permission to use the data of some nonlinear groundwater time series.

## Edit 04-03-2022
Example notebook can be found on [pastas/test-datasets/nonlinear_wrij](https://github.com/pastas/test-datasets/tree/main/nonlinear_wrij)

# Checklist before PR can be merged:
- [x] is documented
- [x] PEP8 compliant code
- [x] tests added / passed
- [x] Example Notebook (for new features)
- [x] API changes documented in Release Notes

### CODE REVIEW

File: pastas/recharge.py [hunk 1]
```diff
@@ -345,14 +354,14 @@ def get_root_zone_balance(pe, ep, srmax=250.0, lp=0.25, ks=100.0,
         """
         n = pe.size
         # Create empty arrays to store the fluxes and states
-        sr = zeros(n + 1, dtype=float64)  # Root Zone Storage State
```
Comment (raoulcollenteur) [inline pastas/recharge.py:348]: This needs to be updated. I think the changes for this PR were made on an old version of the dev-branch. This change is unnecessary and will cause water balance errors at the final time step.

File: pastas/recharge.py [hunk 2]
```diff
@@ -345,14 +354,14 @@ def get_root_zone_balance(pe, ep, srmax=250.0, lp=0.25, ks=100.0,
...
+        sr = zeros(n, dtype=float64)  # Root Zone Storage State
         sr[0] = 0.5 * srmax  # Set the initial system state to half-full
         ea = zeros(n, dtype=float64)  # Actual evaporation Flux
         r = zeros(n, dtype=float64)  # Recharge Flux
         q = zeros(n, dtype=float64)  # Surface runoff Flux
         lp = lp * srmax  # Do this here outside the for-loop for efficiency
 
-        for t in range(n):
+        for t in range(n - 1):
```
Comment (raoulcollenteur) [inline pastas/recharge.py:364]: Unnecessary, remove.

File: pastas/recharge.py [hunk 3]
```diff
@@ -371,7 +380,7 @@ def get_root_zone_balance(pe, ep, srmax=250.0, lp=0.25, ks=100.0,
             # Update storage in the root zone
             sr[t + 1] = sr[t] + dt * (pe[t] - r[t] - ea[t])
 
-        return sr[:-1], -r, -ea, -q, pe
```
Comment (raoulcollenteur) [inline pastas/recharge.py:374]: Unnecessary, remove.

File: pastas/recharge.py [hunk 4]
```diff
@@ -412,19 +421,22 @@ def get_interception_balance(pr, ep, simax=2.0, dt=1.0):
 
         """
         n = pr.size
-        si = zeros(n + 1, dtype=float64)  # Interception Storage State
```
Comment (raoulcollenteur) [inline pastas/recharge.py:415]: Unnecessary, remove.

File: pastas/recharge.py [hunk 5]
```diff
@@ -412,19 +421,22 @@ def get_interception_balance(pr, ep, simax=2.0, dt=1.0):
...
+            # Interception bucket
+            pe[t] = max(pr[t] - simax + si[t], 0.0)
             ei[t] = min(ep[t], si[t])
-            si[t + 1] = si[t] + dt * (pr[t] - ei[t])
-            pe[t] = max(si[t + 1] - simax, 0.0)
-            si[t + 1] = si[t + 1] - pe[t]
+            si[t + 1] = si[t] + dt * (pr[t] - pe[t] - ei[t])
+
+        pe[t + 1] = max(pr[t + 1] - simax + si[t + 1], 0.0)
+        ei[t + 1] = min(ep[t + 1], si[t + 1])
```
Comment (raoulcollenteur) [inline pastas/recharge.py:435]: Unnecessary, remove.

File: pastas/recharge.py [hunk 6]
```diff
@@ -412,19 +421,22 @@ def get_interception_balance(pr, ep, simax=2.0, dt=1.0):
...
 
         pi = pr - pe  # Compute intercepted precipitation
 
-        return si[:-1], -ei, pi
+        return si, -ei, pi
```
Comment (raoulcollenteur) [inline pastas/recharge.py:439]: Unnecessary, remove.

File: pastas/recharge.py [hunk 7]
```diff
@@ -463,18 +475,18 @@ def get_snow_balance(prec, temp, tt=0.0, k=2.0):
         """
         n = prec.size
         # Create empty arrays to store the fluxes and states
-        ss = zeros(n + 1, dtype=float64)  # Snow Storage
+        ss = zeros(n, dtype=float64)  # Snow Storage
```
Comment (raoulcollenteur) [inline pastas/recharge.py:478]: Unnecessary, remove.

File: pastas/recharge.py [hunk 8]
```diff
@@ -463,18 +475,18 @@ def get_snow_balance(prec, temp, tt=0.0, k=2.0):
...
         n = prec.size
         # Create empty arrays to store the fluxes and states
-        ss = zeros(n + 1, dtype=float64)  # Snow Storage
+        ss = zeros(n, dtype=float64)  # Snow Storage
         ps = where(temp <= tt, prec, 0.0)  # Snowfall
         m = where(temp > tt, k * (temp - tt), 0.0)  # Potential Snowmelt
 
         # Snow bucket
-        for t in range(n):
+        for t in range(n - 1):
```
Comment (raoulcollenteur) [inline pastas/recharge.py:483]: Unnecessary, remove.

File: pastas/recharge.py [hunk 9]
```diff
@@ -463,18 +475,18 @@ def get_snow_balance(prec, temp, tt=0.0, k=2.0):
...
             if temp[t] > tt:
                 smoothing_factor = 1.0 - exp(-(ss[t] / 1.5))
                 m[t] = min(m[t] * smoothing_factor, ss[t])
             ss[t + 1] = ss[t] + ps[t] - m[t]
 
-        return ss[:-1], ps, -m
+        return ss, ps, -m
```
Comment (raoulcollenteur) [inline pastas/recharge.py:489]: Unnecessary, remove.

File: pastas/recharge.py [hunk 10]
```diff
@@ -287,8 +292,12 @@ def simulate(self, prec, evap, temp, p, dt=1.0, return_full=False,
 
         if self.gw_uptake:
             # Compute leftover potential evaporation
+            if self.interception:
+                p_f = p[6]
+            else:
+                p_f = p[5]
             eg = ep + ea  # positive flux
-            r = r + eg  #
+            r = r + p_f * eg
```
Comment (raoulcollenteur) [inline pastas/recharge.py:300]: This and the line above should be the only changes to the FlexModel.

File: pastas/recharge.py [hunk 11]
```diff
@@ -221,6 +224,8 @@ def get_init_parameters(self, name="recharge"):
         parameters.loc[name + "_kv"] = (1.0, 0.25, 2.0, False, name)
         if self.interception:
             parameters.loc[name + "_simax"] = (2.0, 0.0, 10.0, False, name)
+        if self.gw_uptake:
+            parameters.loc[name + "_f"] = (1.0, 0.0, 1.0, True, name)
```
Comment (raoulcollenteur) [inline pastas/recharge.py:228]: OK

File: pastas/recharge.py [hunk 12]
```diff
@@ -205,6 +206,8 @@ def __init__(self, interception=True, snow=False, gw_uptake=False):
         self.nparam = 5
         if self.interception:
             self.nparam += 1
+        if self.gw_uptake:
+            self.nparam += 1
```
Comment (raoulcollenteur) [inline pastas/recharge.py:210]: OK

File: pastas/recharge.py [hunk 13]
```diff
@@ -38,7 +38,7 @@
 
 from logging import getLogger
 from numpy import add, float64, multiply, exp, zeros, nan_to_num, vstack, \
-    where
+    where, power
```
Comment (raoulcollenteur) [inline pastas/recharge.py:41]: why not x ** y?

File: pastas/recharge.py [hunk 14]
```diff
@@ -628,3 +640,98 @@ def get_water_balance(self, prec, evap, p, dt=1.0, **kwargs):
         data = DataFrame(data=vstack((s, pe, ea, r)).T,
                          columns=["S", "Pe", "Ea", "R"])
         return data
+
+
+class Peterson(RechargeBase):
+    """Recharge to the groundwater calculated based on [peterson_2014]_.
```
Comment (raoulcollenteur) [inline pastas/recharge.py:646]: Proper doctoring required here.

**Review by raoulcollenteur — APPROVED** [review]

**Comment by raoulcollenteur** [comment]
Thanks, interesting addition. Curious how identifiable the gw evaporation factor is in practice. Couple of things before we can Pull:

- Remove all unnecessary changes in the FlexModel (they cause water balance errors in the final time step).
- Add a test for FlexModel (ps.rch.FlexModel(gw_uptake=True)
- Add tests for Peterson (look at / copy the tests for the FlexModel in the `tests` folder).
- Add docstrings for the Peterson model. Mention you use explicit Euler where Peterson basically uses implicit.
- Should we import Peterson model by default ?
- We could add a code-block to the recharge notebook with this model (where we compare Berendrecht, Flex and Linear already).

**Review by raoulcollenteur — CHANGES_REQUESTED** [review]
See comments in the code and the GH Conversation

File: pastas/recharge.py (hunk 13 above)
Comment (martinvonk) [reply pastas/recharge.py:41]: I believe this is faster with Numba

File: pastas/recharge.py (hunk 1 above)
Comment (martinvonk) [reply pastas/recharge.py:348]: This is weird. On my local (up to date) unsat branch I have the newest version of the Flex Model. But in this PR it shows as if I overwrite these changes.

Comment (martinvonk) [reply pastas/recharge.py:348]: Nevermind. I'll fix it.

**Comment by martinvonk** [comment]
> Thanks, interesting addition. Curious how identifiable the gw evaporation factor is in practice. Couple of things before we can Pull:
> 
> * Remove all unnecessary changes in the FlexModel (they cause water balance errors in the final time step).
> * Add a test for FlexModel (ps.rch.FlexModel(gw_uptake=True)
> * Add tests for Peterson (look at / copy the tests for the FlexModel in the `tests` folder).
> * Add docstrings for the Peterson model. Mention you use explicit Euler where Peterson basically uses implicit.
> * Should we import Peterson model by default ?
> * We could add a code-block to the recharge notebook with this model (where we compare Berendrecht, Flex and Linear already).
> * **Edit**: Open PR That proposes this changes such that it will show up in the release notes automatically :-)

All done. 
The changes to the FlexModel were due to some merge conflicts. I approved the wrong changes, my bad.

Tests are now added for Peterson (except water balance test). I don't think Peterson should be default (yet). Some significant water balance errors might occur still because of using forward Euler. Currently a more quick & dirty implementation is used but this could/should be improved in the future.

**Review by raoulcollenteur — APPROVED** [review]

**Comment by raoulcollenteur** [comment]
Going to check the notebook now and will merge after if all works.

**Pull request merged**